pkg/option/option.go

```diff
@@ -115,200 +120,224 @@ func (om OptionMap) DeepCopy() OptionMap {
 	return cpy
 }
 
-// BoolOptions member functions with external access do not require
+// IntOptions member functions with external access do not require
 // locking by the caller, while functions with internal access presume
 // the caller to have taken care of any locking needed.
-type BoolOptions struct {
+type IntOptions struct {
 	optsMU  lock.RWMutex   // Protects all variables from this structure below this line
 	Opts    OptionMap      `json:"map"`
 	Library *OptionLibrary `json:"-"`
 }
 
 // GetImmutableModel returns the set of immutable options as a ConfigurationMap API model.
-func (bo *BoolOptions) GetImmutableModel() *models.ConfigurationMap {
+func (io *IntOptions) GetImmutableModel() *models.ConfigurationMap {
 	immutableCfg := make(models.ConfigurationMap)
 	return &immutableCfg
 }
 
 // GetMutableModel returns the set of mutable options as a ConfigurationMap API model.
-func (bo *BoolOptions) GetMutableModel() *models.ConfigurationMap {
+func (io *IntOptions) GetMutableModel() *models.ConfigurationMap {
 	mutableCfg := make(models.ConfigurationMap)
-	bo.optsMU.RLock()
-	for k, v := range bo.Opts {
-		if v {
-			mutableCfg[k] = "Enabled"
+	io.optsMU.RLock()
+	for k, v := range io.Opts {
+		if v == OptionDisabled {
+			mutableCfg[k] = fmt.Sprintf("Disabled")
 		} else {
-			mutableCfg[k] = "Disabled"
+			mutableCfg[k] = fmt.Sprintf("Enabled")
 		}
 	}
-	bo.optsMU.RUnlock()
+	io.optsMU.RUnlock()
 
 	return &mutableCfg
 }
 
-func (bo *BoolOptions) DeepCopy() *BoolOptions {
-	bo.optsMU.RLock()
-	cpy := &BoolOptions{
-		Opts:    bo.Opts.DeepCopy(),
-		Library: bo.Library,
+func (io *IntOptions) DeepCopy() *IntOptions {
+	io.optsMU.RLock()
+	cpy := &IntOptions{
+		Opts:    io.Opts.DeepCopy(),
+		Library: io.Library,
 	}
-	bo.optsMU.RUnlock()
+	io.optsMU.RUnlock()
 	return cpy
 }
 
-func NewBoolOptions(lib *OptionLibrary) *BoolOptions {
-	return &BoolOptions{
+func NewIntOptions(lib *OptionLibrary) *IntOptions {
+	return &IntOptions{
 		Opts:    OptionMap{},
 		Library: lib,
 	}
 }
 
-func (bo *BoolOptions) isEnabled(key string) bool {
-	set, exists := bo.Opts[key]
-	return exists && set
+func (io *IntOptions) GetValue(key string) int {
+	value, exists := io.Opts[key]
+	if !exists {
+		return OptionDisabled
+	}
+	return value
+}
+
+func (io *IntOptions) IsEnabled(key string) bool {
+	io.optsMU.RLock()
+	defer io.optsMU.RUnlock()
+	return io.GetValue(key) != OptionDisabled
 }
 
-func (bo *BoolOptions) IsEnabled(key string) bool {
-	bo.optsMU.RLock()
-	defer bo.optsMU.RUnlock()
-	return bo.isEnabled(key)
+// SetValidated sets the option `key` to the specified value. The caller is
+// expected to have validated the input to this function.
+func (io *IntOptions) SetValidated(key string, value int) {
+	io.optsMU.Lock()
```

Comment:
Why not have this validate the key and value regardless? There can be a private function which can assume that validation is done which can be called internally within `pkg/option`. I dislike assuming that the caller has done anything, as it means the code is more prone to misuse.

Reply:
Because sometimes you have pre-validated values and there's no point in validating things twice. For example, this function is called from the Endpoint package where the values are taken directly from the default definitions from the option package. 

Most of the usages of `Set()`, I've converted over to the new `SetBool()`, then I've renamed `Set()` to `SetValidated()` to make this more explicit, so the developer really can't ignore the fact that they're passing a value to the function that assumes the input was validated. I think that it's fine to assume particular behaviour from a caller so long as you make it very explicit, and this change makes it more explicit than it was previously.

We can always make more improvements to this, but I'll focus my effort on getting the functionality of this PR working first.